feat: fp32 vector to fp16/bf16 vector conversion for RESTful API

jiangyinzuo opened the pull request:
RESTful API. The influenced API are as follows:

- Handler. insert
- HandlerV1. insert/upsert
- HandlerV2. insert/upsert/search

We do not modify search API in Handler/HandlerV1 because they do not support fp16/bf16 vectors.

module github.com/milvus-io/milvus/pkg:

Add `Float32ArrayToBFloat16Bytes()`, `Float32ArrayToFloat16Bytes()` and
`Float32ArrayToBytes()`. These method will be used in GoSDK in the
future.

issue: #37448

jiangyinzuo commented on internal/distributed/proxy/httpserver/wrap_request.go:
```diff
@@ -320,18 +367,6 @@ func vector2Bytes(vectors [][]float32) []byte {
-	}
-	return data
-}
-
```
duplicate function with `serialize`, so I removed it.

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo commented:
/run-cpu-e2e

smellthemoon commented on internal/distributed/proxy/httpserver/utils.go:
```diff
@@ -809,9 +805,25 @@ func anyToColumns(rows []map[string]interface{}, validDataMap map[string][]bool,
+					vec := typeutil.Float32ArrayToBFloat16Bytes(candi.v.Interface().([]float32))
+					nameColumns[field.Name] = append(nameColumns[field.Name].([][]byte), vec)
+				default:
+					return nil, fmt.Errorf("invalid type(%v) of field(%v) ", field.DataType, field.Name)
```
same here, please use merr to return the error.

smellthemoon commented on internal/distributed/proxy/httpserver/utils.go:
```diff
@@ -1091,6 +1092,26 @@ func serializeByteVectors(vectorStr string, dataType schemapb.DataType, dimensio
+		values := make([][]byte, 0)
+		for _, vectorArray := range fp32Values {
+			if int64(len(vectorArray)) != dimension {
+				vecStr, _ := json.MarshalToString(vectorArray)
```
the vecStr seems only used to report err? It is recommended not to use marshal one more time to get vecStr, but to pass in other valid error information, such as field name, to prompt the user to pass in the correct data.

jiangyinzuo commented:
/run-cpu-e2e

smellthemoon commented:
/lgtm

zhuwenxing commented:
/lgtm

jiangyinzuo commented:
rerun go-sdk

jiangyinzuo commented:
/run-cpu-e2e

smellthemoon commented:
/lgtm

cqy123456 commented:
/lgtm

czs007 commented on internal/distributed/proxy/httpserver/utils.go:
```diff
@@ -367,27 +365,25 @@ func checkAndSetData(body string, collSchema *schemapb.CollectionSchema) (error,
+					// unmarshal []float32 first to make sure `[3, 3]` is []float instead of []byte
+					var float32Array []float32
+					err := json.Unmarshal([]byte(vectorStr), &float32Array)
+					if err == nil {
```
Is there a chance of a false positive occurring here?

czs007 commented on internal/distributed/proxy/httpserver/utils.go:
```diff
@@ -1118,6 +1119,25 @@ func serializeByteVectors(vectorStr string, dataType schemapb.DataType, dimensio
+	// try to unmarshal as [][]float32 first to make sure `[[3, 3]]` is [][]float32 instead of [][]byte
+	fp32Values := make([][]float32, 0)
+	err := json.Unmarshal([]byte(vectorStr), &fp32Values)
+	if err == nil {
```
Is there a chance of a false positive occurring here?
the performance here is likely to deteriorate, isn't it?

jiangyinzuo commented:
```bash
go test -bench=BenchmarkSerialize -run=^$
```

```
goos: linux
goarch: amd64
pkg: github.com/milvus-io/milvus/internal/distributed/proxy/httpserver
cpu: Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz
BenchmarkSerialize_FloatVectors_Baseline-40                            7         144641969 ns/op
BenchmarkSerialize_FloatVectors-40                                    10         108097462 ns/op
BenchmarkSerialize_FloatVectors_Float16-40                             7         148711147 ns/op
BenchmarkSerialize_FloatOrByteVectors_Fp32-40                          7         158311570 ns/op
BenchmarkSerialize_FloatOrByteVectors_Byte-40                         58          19690757 ns/op
BenchmarkSerialize_FloatOrByteVectors_Fp32_UnmashalTwice-40            7         151955889 ns/op
BenchmarkSerialize_FloatOrByteVectors_Byte_UnmashalTwice-40           37          33288434 ns/op
BenchmarkSerialize_ByteVectors-40                                     55          21864953 ns/op
PASS
ok      github.com/milvus-io/milvus/internal/distributed/proxy/httpserver       17.815s
```

jiangyinzuo committed:
feat: fp32 to fp16/bf16 vector conversion for RESTful API

Make the Milvus server support the conversion from fp32 to fp16/bf16,
so as to facilitate the client users who have difficulties in handling
fp16/bf16 conversions.

1. The influenced RESTful API are as follows:
    - Handler. insert
    - HandlerV1. insert/upsert
    - HandlerV2. insert/upsert/search

We do not modify search API in Handler/HandlerV1 because they do not
support fp16/bf16 vectors.

2. Add micro-benchmarks for various vector serialization methods.
3. Module github.com/milvus-io/milvus/pkg.

Add `Float32ArrayToBFloat16Bytes()`, `Float32ArrayToFloat16Bytes()` and
`Float32ArrayToBytes()`. These method will be used in GoSDK in the
future.

issue: #37448

Signed-off-by: Yinzuo Jiang <yinzuo.jiang@zilliz.com>
Signed-off-by: Yinzuo Jiang <jiangyinzuo@foxmail.com>

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo commented:
rerun cpp-unit-test

jiangyinzuo commented:
rerun go-sdk

czs007 commented:
/approve
/lgtm
